QuickSort Algorithm, Sample Array, Restart Swaps

reeichert opened the pull request:
New Algorithm
- Created QuickSort algorithm

Feature: Restart Swaps
- new dataSourceBackup, to store the original dataSource
- new restart function, that restore dataSource then generate and start all swaps after 1 second.
- handle the restart button after end all swaps

- add new parameter to AlgorithmDetailView protocol, to handle if it need to reload the collection view or not.

Feature: Random Sample Array
- New button to allow user to use a Random sample array.

reeichert committed:
New Algorithm
- Created QuickSort algorithm

Feature: Restart Swaps
- new dataSourceBackup, to store the original dataSource
- new restart function, that restore dataSource then generate and start all swaps after 1 second.
- handle the restart button after end all swaps

- add new parameter to AlgorithmDetailView protocol, to handle if it need to reload the collection view or not.

Feature: Random Sample Array
- New button to allow user to use a Random sample array.

reeichert commented:
@victorpanitz please review

victorpanitz commented on SortingAlgorithms/Scenes/AlgorithmDetail/AlgorithmDetailViewController.swift:
```diff
@@ -83,16 +98,29 @@ final class AlgorithmDetailViewController: UIViewController {
     }
+    
+    @objc private func restartButtonDidTouchUpInside() {
+        presenter.restart()
```
Aqui seria interessante não dar inteligência para a View, no quase quando algo for delegado para a Presenter, delegar de modo que a presenter receba o "acontecimento" da view e nâo uma ordem ou ação. 

Sugestão:` presenter.restartButtonTriggered()`

victorpanitz commented on SortingAlgorithms/Scenes/AlgorithmDetail/AlgorithmDetailViewController.swift:
```diff
@@ -83,16 +98,29 @@ final class AlgorithmDetailViewController: UIViewController {
+    
+    @objc private func restartButtonDidTouchUpInside() {
+        presenter.restart()
+        restartButton.isEnabled = false
```
Sugiro ter um método explicito de enable e de disable do botão no protocolo de view para que os testes possam ser consistentes e essas ações possam ser decididas pelo presenter, invés da view. Desde modo, a view está tomando decisões sobre seus estados.

victorpanitz commented on SortingAlgorithms/Scenes/AlgorithmDetail/AlgorithmDetailViewController.swift:
```diff
@@ -83,16 +98,29 @@ final class AlgorithmDetailViewController: UIViewController {
 
 extension AlgorithmDetailViewController: AlgorithmDetailView {
     
+    func allSwapsDidEnd() {
```
Sendo o protocolo de view o meio da presenter executar sua tomada de decisão sobre a regra de apresentação, acho que esse método poderia ser renomeado pára algo que represente o que ele esteja executando, como por exemplo `enableRestartButton()`.

victorpanitz commented on SortingAlgorithms/Scenes/AlgorithmDetail/AlgorithmDetailViewController.swift:
```diff
@@ -83,16 +98,29 @@ final class AlgorithmDetailViewController: UIViewController {
     }
     
-    func updateDataSource(_ datasource: [Int]) {
+    func updateDataSource(_ datasource: [Int], reloadData: Bool) {
```
A lógica de apresentação deve ficar no lado da Presenter, logo valeria rever a necessidade de atualizar o datasource sem a necessidade de fazer o `reloadData`, caso necessário atualizar o dataSource mesmo sem o reloadData, vale criar um método explicito de reloadData para que a Presenter tome essa decisão e fique tranquilo para testar.

victorpanitz commented on SortingAlgorithms/Scenes/DataSourceInput/DataSourceViewController.swift:
```diff
@@ -161,6 +187,10 @@ final class DataSourceViewController: UIViewController {
     }
+    
+    @objc private func addSampleArrayTouchUpInside() {
+        presenter.addSampleArray()
```
Mesma questão da delegação de um acontecimento na view, invés de passar uma ordem, passar um evento e.g `presenter.addSampleButtonTrigerred()`

victorpanitz reviewed (commented):
Muito massa meu, depois que mergear, já vou atualizar o README, adicionar os testes da camada de apresentação e lançar uma nova release pra loja ;) Mandou bem

reeichert commented:
Nice! vou fazer essas fixes e te retorno!

reeichert committed:
- Update QuickSort Algorithm
- Update Presenter and View Protocols
- Added QuickSort Tests
- Update DetailPresenter and ListPresenter tests

reeichert commented:
@victorpanitz Adicionei testes e fiz os fixes necessários.

victorpanitz reviewed (approved):
Massa mano!!
